server/src/main/java/org/opensearch/action/search/CreatePitController.java

```diff
@@ -0,0 +1,268 @@
+    private final CreatePitRequest request;
+    private static final Logger logger = LogManager.getLogger(CreatePitController.class);
+    public static final Setting<TimeValue> CREATE_PIT_TEMPORARY_KEEPALIVE_SETTING = Setting.positiveTimeSetting(
+        "pit.temporary.keep_alive_interval",
+        timeValueSeconds(30),
+        Setting.Property.NodeScope
+    );
+
+    public CreatePitController(
+        CreatePitRequest request,
+        SearchTransportService searchTransportService,
+        ClusterService clusterService,
+        TransportSearchAction transportSearchAction,
+        NamedWriteableRegistry namedWriteableRegistry,
+        Task task,
+        ActionListener<CreatePitResponse> listener
+    ) {
+        this.searchTransportService = searchTransportService;
+        this.clusterService = clusterService;
+        this.transportSearchAction = transportSearchAction;
+        this.namedWriteableRegistry = namedWriteableRegistry;
+        this.task = task;
+        this.listener = listener;
+        this.request = request;
+        runner = this::executeCreatePit;
+    }
+
+    private TimeValue getCreatePitTemporaryKeepAlive() {
+        return CREATE_PIT_TEMPORARY_KEEPALIVE_SETTING.get(clusterService.getSettings());
+    }
+
+    /**
+     * Method for creating PIT reader context
+     * Phase 1 of create PIT request : Create PIT reader contexts in the associated shards with a temporary keep alive
+     * Phase 2 of create PIT : Update PIT reader context with PIT ID and keep alive from request and
+     * fail user request if any of the updates in this phase are failed - we clean up PITs in case of such failures
+     */
+    public void executeCreatePit() {
+        SearchRequest searchRequest = new SearchRequest(request.getIndices());
+        searchRequest.preference(request.getPreference());
+        searchRequest.routing(request.getRouting());
+        searchRequest.indicesOptions(request.getIndicesOptions());
+        searchRequest.allowPartialSearchResults(request.shouldAllowPartialPitCreation());
+
+        SearchTask searchTask = new SearchTask(
+            task.getId(),
+            task.getType(),
+            task.getAction(),
+            () -> task.getDescription(),
+            task.getParentTaskId(),
+            new HashMap<>()
+        );
+
+        final StepListener<SearchResponse> createPitListener = new StepListener<>();
+
+        final ActionListener<CreatePitResponse> updatePitIdListener = ActionListener.wrap(r -> listener.onResponse(r), e -> {
+            logger.error("PIT creation failed while updating PIT ID", e);
+            listener.onFailure(e);
+        });
+        /**
+         * Phase 1 of create PIT
+         */
+        executeCreatePit(searchTask, searchRequest, createPitListener);
+
+        /**
+         * Phase 2 of create PIT where we update pit id in pit contexts
+         */
+        executeUpdatePitId(request, createPitListener, updatePitIdListener);
+    }
+
```

Comment:
I would re-write this as 
```
final StepListener<SearchResponse> createPitListener = new StepListener<>();
//  Phase 1 of create PIT
executeCreatePit(searchTask, searchRequest, createPitListener);
//  Phase 2 of create PIT
createPitListener.whenComplete(searchResponse -> executeUpdatePitId(request, createPitListener, updatePitIdListener, updatePitIdListener::onFailure)


```

Reply:
Lets pass the `StepListener`  to the method invocation which can aid with stubbing failures